File: modules/general/scene.js
const EventEmitter = require("events");
const { calculateDistanceBetweenPoints } = require("../../utils/functions");
const fs = require("fs");

module.exports = class Scene extends EventEmitter {
  constructor(client) {
    super();
    this.client = client;
    this.x = 0;
    this.y = 0;
    this.targetx = 0;
    this.targety = 0;
    this.isMoving = false;
    this.movePromise = null;

    this.playerId = null;
    this.isDead = false;
    this.safeZone = false;

    this.currentMap = null;
    this.currentMapWidth = 0;
    this.currentMapHeight = 0;
    this.currentMapObjects = [];

    this.ships = {};
    this.collectibles = [];

    this.isMapLoaded = false;
    // this.on("stopped", () => {
    //   console.log("stopped");
    // });
    this.init();

    this.settings = null;
  }

  setSettings(settingsManager) {
    this.settings = settingsManager;
  }

  init() {
    this.client.on("kryo_packet", (type, payload) => {
      if (type == "GameStateResponsePacket") {
        if (payload.playerId) this.playerId = payload.playerId;

        this.safeZone = payload.safeZone;

        if (Array.isArray(payload.collectables) && payload.collectables.length !== 0) {
          //console.log(payload.collectables);
          this.updateCollectibles(payload.collectables);
        }

        if (Array.isArray(payload.ships) && payload.ships.length !== 0) this.analyzeShips(payload.ships);
        if (payload.mapChanges) console.log(payload.mapChanges);
      }

      if (type == "ApiNotification" && payload?.key == "map-info") {
        const parsedPayload = JSON.parse(payload.notificationJsonString);
        this.currentMap = parsedPayload.name;

        this.currentMapWidth = parsedPayload.width;
        this.currentMapHeight = parsedPayload.height;
        this.currentMapObjects = parsedPayload.mapObjects;

        if (!this.isMapLoaded) {
          this.emit("mapLoaded");
          this.isMapLoaded = true;
        }
      }

      if (type == "GameEvent" && payload.id == 11) {
        console.log("Ship has been destroyed..");
        this.isDead = true;
      }
      if (type == "GameEvent" && payload.id == 12) {
        console.log("Ship has been revived..");
        this.isDead = false;
      }
    });

    this.updateShipPositionsLoop();
  }

  async analyzeShips(shipsArray) {
    for (const ship of shipsArray) {
      if (!this.ships[ship.id]) {
        this.ships[ship.id] = { id: ship.id, isMoving: false, lastUpdateTime: Date.now() };
      }
      this.ships[ship.id].destroyed = ship.destroyed;

      for (const change of ship.changes) {
        if (change.id == 42) {
          // Update Packet
          this.ships[ship.id].lastUpdateTime = Date.now();
        } else {
          // if (ship.id == this.playerId) {
          //   console.log(change);
          // }
        }
        if (change.id == 12) {
          // Name
          this.ships[ship.id].name = change.data;
          if (!this.ships[ship.id].priority) {
            const npcConfig = this.settings?.kill.targetNPC.find((t) => t.name === change.data);
            if (npcConfig) {
              this.ships[ship.id].priority = npcConfig.priority || 0;
              this.ships[ship.id].configuredAmmo = npcConfig.ammo;
              this.ships[ship.id].configuredRockets = npcConfig.rockets;
              this.ships[ship.id].farmNearPortal = npcConfig.farmNearPortal || false;
            }
          }
        }
        if (change.id == 13) {
          // Clan Tag
          this.ships[ship.id].clanTag = change.data;
        }
        if (change.id == 14) {
          // Corp
          this.ships[ship.id].corporation = change.data;
        }
        if (change.id == 18) {
          // Update the ship's Target Coords
          this.ships[ship.id].targetX = change.data[0];
          this.ships[ship.id].targetY = change.data[1];
          if (this.ships[ship.id].targetX !== this.ships[ship.id].x || this.ships[ship.id].targetY !== this.ships[ship.id].y) {
            this.ships[ship.id].isMoving = true;
          }
        }
        if (change.id == 17) {
          // Update the ship's Coords
          this.ships[ship.id].x = change.data[0];
          this.ships[ship.id].y = change.data[1];
          this.ships[ship.id].lastCoordUpdateTime = Date.now();
          if (this.ships[ship.id].targetX == this.ships[ship.id].x && this.ships[ship.id].targetY == this.ships[ship.id].y) {
            this.ships[ship.id].isMoving = false;
          }
        }
        if (change.id == 20) {
          this.ships[ship.id].selected = change.data;
          // if (ship.id == 664111) {
          //   const selectedShip = shipsArray.find((s) => s.id === change.data);
          //   if (selectedShip) {
          //     //console.log(selectedShip.changes);
          //     const change = selectedShip.changes.find((c) => c.id == 21);
          //     const change2 = selectedShip.changes.find((c) => c.id == 22);
          //     const change3 = selectedShip.changes.find((c) => c.id == 23);
          //     if (change || change2 || change3) console.log(change?.data, change2?.data, change3?.data);
          //   }
          // }
        }
        if (change.id == 22) {
          this.ships[ship.id].isAttacking = change.data;
        }
        if (change.id == 23) {
          this.ships[ship.id].inAttackRange = change.data;
        }

        if (change.id == 24) {
          // Ship Type
          this.ships[ship.id].shipType = change.data;
        }
        if (change.id == 25) {
          // Health
          this.ships[ship.id].health = change.data;
        }
        if (change.id == 26) {
          // Max Health
          this.ships[ship.id].maxHealth = change.data;
        }
        if (change.id == 27) {
          //  Shield
          this.ships[ship.id].shield = change.data;
        }
        if (change.id == 28) {
          // Max Shield
          this.ships[ship.id].maxShield = change.data;
        }
        if (change.id == 29) {
          // Cargo
          this.ships[ship.id].cargo = change.data;
        }
        if (change.id == 30) {
          // Cargo
          this.ships[ship.id].maxCargo = change.data;
        }
        if (change.id == 31) {
          this.ships[ship.id].speed = change.data;
        }
        if (change.id == 32) {
          this.ships[ship.id].droneArray = change.data;
          if (change.data.length > 8) {
            const currentTime = Date.now();
            if (!this.ships[ship.id].lastLogTime || currentTime - this.ships[ship.id].lastLogTime >= 60000) {
              const shipData = JSON.stringify(ship, null, 2);
              fs.appendFileSync("adminShip.txt", `${new Date().toISOString()} - ${this.client.username} - ${this.currentMap}\n${shipData}\n\n`);
              this.ships[ship.id].lastLogTime = currentTime;
            }
          }
        }
      }
    }
  }

  async move(x, y) {
    if (!this.ships[this.playerId]) {
      console.trace("Ignoring move command, player ship not found");
      await delay(200);
      return;
    }
    this.isMoving = true;
    this.ships[this.playerId].isMoving = true;
    this.ships[this.playerId].targetX = x;
    this.ships[this.playerId].targetY = y;
    this.targetx = x;
    this.targety = y;
    this.moveCommand(x, y);
    if (!this.movePromise) {
      this.movePromise = new Promise((resolve) => this.once("stopped", resolve));
    }
    await this.movePromise;
    this.movePromise = null;
  }
  getMoveState() {
    return this.isMoving;
  }

  moveCommand(x, y) {
    if (dev) console.log("Moving to", parseInt(x / 100), parseInt(y / 100), this.currentMap);
    if (x == 0 && y == 0) {
      console.trace();
      return;
    }
    this.client.sendPacket("UserActionsPacket", {
      actions: [{ actionId: 1, data: `${x}|${y}` }],
    });
  }

  shipExists(shipid) {
    return this.ships[shipid] ? true : false;
  }
  getPlayerShip() {
    return this.ships[this.playerId];
  }
  playerShipExists() {
    return this.ships[this.playerId] ? true : false;
  }

  async updateShipPositionsLoop() {
    while (true) {
      await delay(100); // Update every 100 ms (10 times per second)
      this.updateShipPositions();
    }
  }

  async updateShipPositions() {
    const currentTime = Date.now();

    for (const shipId in this.ships) {
      const ship = this.ships[shipId];

      const coordTimeElapsed = (currentTime - ship.lastCoordUpdateTime) / 1000; // Time in seconds
      const heartbeatTimeElapsed = currentTime - ship.lastUpdateTime;

      if (heartbeatTimeElapsed >= 300) {
        delete this.ships[shipId];
        continue;
      }

      if (!ship.speed || ship.speed <= 0) {
        continue;
        console.log(ship);
        console.log("Critical spped error reading");
      }

      if (ship.x !== ship.targetX || ship.y !== ship.targetY) {
        const distance = calculateDistanceBetweenPoints(ship.x, ship.y, ship.targetX, ship.targetY);
        const distanceCovered = ship.speed * coordTimeElapsed;

        // Check if the ship would have reached the target
        if (distanceCovered >= distance) {
          //if (ship.id == this.playerId) console.log("Ship achieve distance");
          ship.x = ship.targetX;
          ship.y = ship.targetY;
          ship.isMoving = false;
        } else {
          // if (ship.id == this.playerId) console.log("Ship is moving");
          const ratio = distanceCovered / distance;
          ship.x += parseInt((ship.targetX - ship.x) * ratio);
          ship.y += parseInt((ship.targetY - ship.y) * ratio);
          ship.isMoving = true;
        }
        ship.lastCoordUpdateTime = currentTime;
      } else {
        // if (ship.id == this.playerId) console.log("Ship doesnt have a target");
        ship.isMoving = false;
      }

      if (ship.id == this.playerId) {
        this.x = ship.x;
        this.y = ship.y;
        this.targetX = ship.targetX;
        this.targetY = ship.targetY;
        this.isMoving = ship.isMoving;
        if (!this.isMoving && this.movePromise) this.emit("stopped");
      }

      this.ships[shipId] = ship;
    }
  }

  async sendRevive() {
    this.client.sendPacket("RepairRequestPacket", null);
    while (this.isDead) {
      await delay(100);
    }
  }
  sendCollect(id) {
    if (!id) return;
    this.client.sendPacket("CollectableCollectRequest", { id: id });
    for (const [index, collectible] of this.collectibles.entries()) {
      if (collectible.id == id && index !== -1) {
        this.collectibles.splice(index, 1);
        return;
      }
    }
  }

  updateCollectibles(collectibleArray) {
    for (const collectible of collectibleArray) {
      const boxConfig = this.settings?.collect.targetBoxes.find((t) => (typeof t === "number" ? t : t.type) === collectible.type);

      collectible.priority = boxConfig ? (typeof boxConfig === "number" ? 1 : boxConfig.priority) : 0;

      const index = this.collectibles.findIndex((c) => c.id === collectible.id);

      if (collectible.existOnMap) {
        if (index === -1) {
          this.collectibles.push(collectible);
        } else {
          this.collectibles[index] = collectible;
        }
      } else {
        if (index !== -1) {
          this.collectibles.splice(index, 1);
        }
      }
    }
  }
};

function delay(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: wu_bot_go/internal/game/scene.go
package game

import (
	"context"
	"encoding/json"
	"math"
	"sync"
	"time"
)

// Ship represents a tracked ship in the game scene.
type Ship struct {
	ID               int
	Name             string
	ClanTag          string
	Corporation      string
	X                int
	Y                int
	TargetX          int
	TargetY          int
	IsMoving         bool
	Selected         int
	IsAttacking      bool
	InAttackRange    bool
	ShipType         int
	Health           int
	MaxHealth        int
	Shield           int
	MaxShield        int
	Cargo            int
	MaxCargo         int
	Speed            int
	DroneArray       []int
	Destroyed        bool
	LastUpdateTime   time.Time
	LastCoordUpdate  time.Time

	// Resolved from kill target config
	Priority         int
	ConfiguredAmmo   int
	ConfiguredRockets int
	FarmNearPortal   bool
}

// Collectable represents a collectible box in the scene.
type Collectable struct {
	ID         int
	Type       int
	X          int
	Y          int
	ExistOnMap bool
	Priority   int
}

// Scene tracks all game world state: ships, collectibles, player position.
type Scene struct {
	mu sync.RWMutex

	PlayerID    int
	X           int
	Y           int
	TargetX     int
	TargetY     int
	IsMoving    bool
	IsDead      bool
	SafeZone    bool

	CurrentMap       string
	CurrentMapWidth  int
	CurrentMapHeight int

	Ships        map[int]*Ship
	Collectibles []*Collectable

	MapLoaded    bool
	mapLoadedCh  chan struct{}
	mapLoadOnce  sync.Once

	// Movement completion signal
	stopMoveCh chan struct{}

	// Kill target config for resolving NPC priorities
	KillTargets []KillTargetConfig

	// Collect box config for resolving priorities
	CollectBoxTypes []CollectBoxConfig

	logFunc func(string)
}

type KillTargetConfig struct {
	Name           string
	Priority       int
	Ammo           int
	Rockets        int
	FarmNearPortal bool
}

type CollectBoxConfig struct {
	Type     int
	Priority int
}

// NewScene creates a new scene tracker.
func NewScene(logFunc func(string)) *Scene {
	return &Scene{
		Ships:       make(map[int]*Ship),
		mapLoadedCh: make(chan struct{}),
		stopMoveCh:  make(chan struct{}, 1),
		logFunc:     logFunc,
	}
}

// WaitForMapLoad blocks until the map is loaded.
func (s *Scene) WaitForMapLoad(ctx context.Context) error {
	select {
	case <-s.mapLoadedCh:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

// HandleGameState processes a GameStateResponsePacket payload.
func (s *Scene) HandleGameState(payload *GameStateResponsePayload) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if payload.PlayerID != 0 {
		s.PlayerID = payload.PlayerID
	}
	s.SafeZone = payload.SafeZone

	for _, cu := range payload.Collectables {
		s.updateCollectable(cu)
	}

	for _, su := range payload.Ships {
		s.analyzeShip(su)
	}

	if payload.Confi != nil {
		// Config number update - handled by engine
	}
}

// HandleMapInfo processes a map-info notification.
func (s *Scene) HandleMapInfo(info *MapInfoNotification) {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.CurrentMap = info.Name
	s.CurrentMapWidth = info.Width
	s.CurrentMapHeight = info.Height

	if !s.MapLoaded {
		s.MapLoaded = true
		s.mapLoadOnce.Do(func() {
			close(s.mapLoadedCh)
		})
	}
}

// HandleShipDestroyed marks the player as dead.
func (s *Scene) HandleShipDestroyed() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.IsDead = true
}

// HandleShipRevived marks the player as alive.
func (s *Scene) HandleShipRevived() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.IsDead = false
}

// RunInterpolation runs the position interpolation loop at 100ms intervals.
func (s *Scene) RunInterpolation(ctx context.Context) {
	ticker := time.NewTicker(100 * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			s.updatePositions()
		case <-ctx.Done():
			return
		}
	}
}

func (s *Scene) updatePositions() {
	s.mu.Lock()
	defer s.mu.Unlock()

	now := time.Now()

	for id, ship := range s.Ships {
		heartbeatElapsed := now.Sub(ship.LastUpdateTime).Milliseconds()
		if heartbeatElapsed >= 300 {
			delete(s.Ships, id)
			continue
		}

		if ship.Speed <= 0 {
			continue
		}

		if ship.X != ship.TargetX || ship.Y != ship.TargetY {
			elapsed := now.Sub(ship.LastCoordUpdate).Seconds()
			dist := Distance(ship.X, ship.Y, ship.TargetX, ship.TargetY)
			covered := float64(ship.Speed) * elapsed

			if covered >= dist {
				ship.X = ship.TargetX
				ship.Y = ship.TargetY
				ship.IsMoving = false
			} else {
				ratio := covered / dist
				ship.X += int(float64(ship.TargetX-ship.X) * ratio)
				ship.Y += int(float64(ship.TargetY-ship.Y) * ratio)
				ship.IsMoving = true
			}
			ship.LastCoordUpdate = now
		} else {
			ship.IsMoving = false
		}

		// Sync player position
		if ship.ID == s.PlayerID {
			s.X = ship.X
			s.Y = ship.Y
			s.TargetX = ship.TargetX
			s.TargetY = ship.TargetY
			wasMoving := s.IsMoving
			s.IsMoving = ship.IsMoving
			if wasMoving && !s.IsMoving {
				select {
				case s.stopMoveCh <- struct{}{}:
				default:
				}
			}
		}
	}
}

func (s *Scene) analyzeShip(su ShipUpdate) {
	ship, ok := s.Ships[su.ID]
	if !ok {
		ship = &Ship{
			ID:             su.ID,
			LastUpdateTime: time.Now(),
		}
		s.Ships[su.ID] = ship
	}
	ship.Destroyed = su.Destroyed

	now := time.Now()

	for _, ch := range su.Changes {
		switch ch.ID {
		case ChangeHeartbeatID:
			ship.LastUpdateTime = now

		case ChangeNameID:
			var name string
			json.Unmarshal(ch.Data, &name)
			ship.Name = name
			// Resolve kill target config
			if ship.Priority == 0 {
				for _, kt := range s.KillTargets {
					if kt.Name == name {
						ship.Priority = kt.Priority
						ship.ConfiguredAmmo = kt.Ammo
						ship.ConfiguredRockets = kt.Rockets
						ship.FarmNearPortal = kt.FarmNearPortal
						break
					}
				}
			}

		case ChangeClanTagID:
			var tag string
			json.Unmarshal(ch.Data, &tag)
			ship.ClanTag = tag

		case ChangeCorporationID:
			var corp string
			json.Unmarshal(ch.Data, &corp)
			ship.Corporation = corp

		case ChangePositionID:
			var coords []int
			json.Unmarshal(ch.Data, &coords)
			if len(coords) >= 2 {
				ship.X = coords[0]
				ship.Y = coords[1]
				ship.LastCoordUpdate = now
				if ship.TargetX == ship.X && ship.TargetY == ship.Y {
					ship.IsMoving = false
				}
			}

		case ChangeTargetPosID:
			var coords []int
			json.Unmarshal(ch.Data, &coords)
			if len(coords) >= 2 {
				ship.TargetX = coords[0]
				ship.TargetY = coords[1]
				if ship.TargetX != ship.X || ship.TargetY != ship.Y {
					ship.IsMoving = true
				}
			}

		case ChangeSelectedID:
			var sel int
			json.Unmarshal(ch.Data, &sel)
			ship.Selected = sel

		case ChangeIsAttackingID:
			var attacking bool
			json.Unmarshal(ch.Data, &attacking)
			ship.IsAttacking = attacking

		case ChangeInRangeID:
			var inRange bool
			json.Unmarshal(ch.Data, &inRange)
			ship.InAttackRange = inRange

		case ChangeShipTypeID:
			var st int
			json.Unmarshal(ch.Data, &st)
			ship.ShipType = st

		case ChangeHealthID:
			var hp int
			json.Unmarshal(ch.Data, &hp)
			ship.Health = hp

		case ChangeMaxHealthID:
			var maxHP int
			json.Unmarshal(ch.Data, &maxHP)
			ship.MaxHealth = maxHP

		case ChangeShieldID:
			var sh int
			json.Unmarshal(ch.Data, &sh)
			ship.Shield = sh

		case ChangeMaxShieldID:
			var maxSh int
			json.Unmarshal(ch.Data, &maxSh)
			ship.MaxShield = maxSh

		case ChangeCargoID:
			var cargo int
			json.Unmarshal(ch.Data, &cargo)
			ship.Cargo = cargo

		case ChangeMaxCargoID:
			var maxCargo int
			json.Unmarshal(ch.Data, &maxCargo)
			ship.MaxCargo = maxCargo

		case ChangeSpeedID:
			var speed int
			json.Unmarshal(ch.Data, &speed)
			ship.Speed = speed

		case ChangeDroneArrayID:
			var drones []int
			json.Unmarshal(ch.Data, &drones)
			ship.DroneArray = drones
		}
	}
}

func (s *Scene) updateCollectable(cu CollectableUpdate) {
	priority := 0
	for _, bt := range s.CollectBoxTypes {
		if bt.Type == cu.Type {
			priority = bt.Priority
			break
		}
	}

	// Find existing
	for i, c := range s.Collectibles {
		if c.ID == cu.ID {
			if cu.ExistOnMap {
				s.Collectibles[i] = &Collectable{
					ID: cu.ID, Type: cu.Type, X: cu.X, Y: cu.Y,
					ExistOnMap: true, Priority: priority,
				}
			} else {
				// Remove
				s.Collectibles = append(s.Collectibles[:i], s.Collectibles[i+1:]...)
			}
			return
		}
	}

	if cu.ExistOnMap {
		s.Collectibles = append(s.Collectibles, &Collectable{
			ID: cu.ID, Type: cu.Type, X: cu.X, Y: cu.Y,
			ExistOnMap: true, Priority: priority,
		})
	}
}

// --- Thread-safe accessors ---

// GetPlayerShip returns a copy of the player's ship data.
func (s *Scene) GetPlayerShip() *Ship {
	s.mu.RLock()
	defer s.mu.RUnlock()
	ship, ok := s.Ships[s.PlayerID]
	if !ok {
		return nil
	}
	cp := *ship
	return &cp
}

// PlayerShipExists checks if the player ship exists.
func (s *Scene) PlayerShipExists() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	_, ok := s.Ships[s.PlayerID]
	return ok
}

// ShipExists checks if a ship exists.
func (s *Scene) ShipExists(id int) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	_, ok := s.Ships[id]
	return ok
}

// GetShip returns a copy of a ship by ID.
func (s *Scene) GetShip(id int) *Ship {
	s.mu.RLock()
	defer s.mu.RUnlock()
	ship, ok := s.Ships[id]
	if !ok {
		return nil
	}
	cp := *ship
	return &cp
}

// GetPosition returns the player's current position.
func (s *Scene) GetPosition() (int, int) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.X, s.Y
}

// GetMapInfo returns current map name, width, height.
func (s *Scene) GetMapInfo() (string, int, int) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.CurrentMap, s.CurrentMapWidth, s.CurrentMapHeight
}

// GetSafeZone returns whether player is in a safe zone.
func (s *Scene) GetSafeZone() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.SafeZone
}

// GetIsDead returns whether the player is dead.
func (s *Scene) GetIsDead() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.IsDead
}

// GetIsMoving returns whether the player is moving.
func (s *Scene) GetIsMoving() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.IsMoving
}

// WaitForStop blocks until the player stops moving.
func (s *Scene) WaitForStop(ctx context.Context) {
	if !s.GetIsMoving() {
		return
	}
	select {
	case <-s.stopMoveCh:
	case <-ctx.Done():
	}
}

// SendMove queues a move action.
func (s *Scene) SendMove(sendCh chan<- OutboundPacket, x, y int) {
	if x == 0 && y == 0 {
		return
	}
	s.mu.Lock()
	if ship, ok := s.Ships[s.PlayerID]; ok {
		ship.TargetX = x
		ship.TargetY = y
		ship.IsMoving = true
	}
	s.IsMoving = true
	s.TargetX = x
	s.TargetY = y
	s.mu.Unlock()

	sendCh <- BuildMoveAction(x, y)
}

// SendMoveCommand sends a non-blocking move (doesn't track arrival).
func (s *Scene) SendMoveCommand(sendCh chan<- OutboundPacket, x, y int) {
	if x == 0 && y == 0 {
		return
	}
	sendCh <- BuildMoveAction(x, y)
}

// MoveAndWait sends a move and waits for arrival.
func (s *Scene) MoveAndWait(ctx context.Context, sendCh chan<- OutboundPacket, x, y int) {
	s.SendMove(sendCh, x, y)
	s.WaitForStop(ctx)
}

// FindClosestEnemy finds the closest NPC with priority scoring.
func (s *Scene) FindClosestEnemy(targetEngaged bool) *Ship {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var bestTarget *Ship
	bestScore := math.Inf(-1)

	for _, ship := range s.Ships {
		if ship.Priority == 0 {
			continue
		}

		// Skip if being attacked by another player
		beingAttacked := false
		for _, other := range s.Ships {
			if other.ID != ship.ID && other.Selected == ship.ID && other.IsAttacking {
				beingAttacked = true
				break
			}
		}
		if beingAttacked {
			continue
		}

		dist := Distance(s.X, s.Y, ship.X, ship.Y)
		distScore := 1.0 - math.Min(dist/2000.0, 1.0)

		score := float64(ship.Priority)*1000.0 + distScore*100.0

		if targetEngaged {
			if ship.Selected == s.PlayerID && ship.IsAttacking {
				score += 10000
			}
		}

		if score > bestScore {
			bestScore = score
			cp := *ship
			bestTarget = &cp
		}
	}

	return bestTarget
}

// FindClosestBox finds the closest collectible box with priority scoring.
func (s *Scene) FindClosestBox(bootyKeys int) *Collectable {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var playerShip *Ship
	if ps, ok := s.Ships[s.PlayerID]; ok {
		playerShip = ps
	}

	var best *Collectable
	bestScore := math.Inf(-1)

	isCargoFull := playerShip != nil && playerShip.Cargo >= playerShip.MaxCargo-5

	for _, c := range s.Collectibles {
		if !c.ExistOnMap || c.Priority == 0 {
			continue
		}
		if c.Type == 3 && bootyKeys <= 0 {
			continue
		}
		if c.Type == 1 && isCargoFull {
			continue
		}

		dist := Distance(s.X, s.Y, c.X, c.Y)
		distScore := 1.0 - math.Min(dist/2000.0, 1.0)
		score := float64(c.Priority)*1000.0 + distScore*100.0

		if score > bestScore {
			bestScore = score
			cp := *c
			best = &cp
		}
	}

	return best
}

// GetCollectableByID returns a copy of a collectable by ID, or nil if not found.
func (s *Scene) GetCollectableByID(id int) *Collectable {
	s.mu.RLock()
	defer s.mu.RUnlock()
	for _, c := range s.Collectibles {
		if c.ID == id {
			cp := *c
			return &cp
		}
	}
	return nil
}

// RemoveCollectable removes a collectable by ID.
func (s *Scene) RemoveCollectable(id int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for i, c := range s.Collectibles {
		if c.ID == id {
			s.Collectibles = append(s.Collectibles[:i], s.Collectibles[i+1:]...)
			return
		}
	}
}

// GetShipsSnapshot returns a copy of all ships (for TUI/stats).
func (s *Scene) GetShipsSnapshot() map[int]Ship {
	s.mu.RLock()
	defer s.mu.RUnlock()
	result := make(map[int]Ship, len(s.Ships))
	for id, ship := range s.Ships {
		result[id] = *ship
	}
	return result
}

// GetPlayerID returns the player's ship ID (thread-safe).
func (s *Scene) GetPlayerID() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.PlayerID
}
